### ISSUE TITLE
Value cannot be opened: "" (Registry)

### ISSUE THREAD

**Issue by blackcomb-dev** [spikespaz/search-deflector#8]
I have encountered an error launching Search Deflector.
The error information follows below.

**File:** `std\windows\registry.d`

**Line:** 495

**Message:**
```
Value cannot be opened: "": The system cannot find the file specified. (error 2)
```

**Stack Trace:**
```
0x00007FF7C53EF0CD
0x00007FF7C53EEE26
0x00007FF7C53E7CDC
0x00007FF7C53D477C
0x00007FF7C53898BA
0x00007FF7C538B708
0x00007FF7C5377076
0x00007FF7C538C4F2
0x00007FF7C538C456
0x00007FF7C5375C32
0x00007FF7C5375AD0
0x00007FF7C53DAD8D
0x00007FF7C53DAC3A
0x00007FF7C53F9658
0x00007FFB89B3FEA4 in BaseThreadInitThunk
0x00007FFB8A985521 in RtlUserThreadStart

```

**Comment by spikespaz** [comment]
@fernandex00 <= Tagging you here since you posted a duplicate.

To both of you; are you running any Antivirus software besides Windows, and what is it?

**Comment by spikespaz** [comment]
@shalvah You said you are experiencing the same issue in the duplicate issue.

Could any of you elaborate? When does this happen, after setup or after making a search?
Are you running any antivirus software besides Windows Defender?

**Comment by shalvah** [comment]
It happens during setup, once the command line window launches.

**Comment by TruckMangione** [comment]
Same issue. Install the program, launcher starts and then immediately fails. Only Windows Defender running, no other antivirus program running.

**Comment by fernandex00** [comment]
Only Windows defender

**Comment by spikespaz** [comment]
@fernandex00 @TruckMangione @shalvah 

Please use `regedit` to retrieve the values from these registry keys. I need all of the values in these areas recursively.

```
HKEY_CURRENT_USER\\Software\\Clients\\SearchDeflector
HKEY_LOCAL_MACHINE\\Software\\Clients\\SearchDeflector
HKEY_LOCAL_MACHINE\\Software\\RegisteredApplications
HKEY_CLASSES_ROOT\\SearchDeflector
```

Also, run the `winver` command and me the full text of the line that has the version. Di you have updates? Are you on 1803?

I can't really think of a cause for this bug, so I'm trying to figure out as much as possible.

**Comment by loginnotsecure** [comment]
@spikespaz no antivirus other than Windows defender. Running Windows build 19H1 build 18214.

**Comment by TruckMangione** [comment]
Version 1803 (os build 17134.228)
No available updates.

Pardon my ignorance, but what information do you want from regedit? Do you want an export of each folder and then for me to attach the reg files here?

**Comment by TruckMangione** [comment]
@spikespaz Three out of four of the keys are coming up blank other than the key itself.

> [HKEY_CURRENT_USER\Software\Clients\SearchDeflector]
> 

> [HKEY_LOCAL_MACHINE\SOFTWARE\Clients\SearchDeflector]
> 

> [HKEY_LOCAL_MACHINE\SOFTWARE\RegisteredApplications]
> "File Explorer"="SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\Capabilities"
> "Paint"="SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Applets\\Paint\\Capabilities"
> "Windows Address Book"="Software\\Clients\\Contacts\\Address Book\\Capabilities"
> "Windows Disc Image Burner"="Software\\Microsoft\\IsoBurn\\Capabilities"
> "Windows Photo Viewer"="Software\\Microsoft\\Windows Photo Viewer\\Capabilities"
> "Windows Search"="Software\\Microsoft\\Windows Search\\Capabilities"
> "Wordpad"="Software\\Microsoft\\Windows\\CurrentVersion\\Applets\\Wordpad\\Capabilities"
> "VLC"="Software\\Clients\\Media\\VLC\\Capabilities"
> "Google Chrome"="Software\\Clients\\StartMenuInternet\\Google Chrome\\Capabilities"
> "WinRAR"="Software\\WinRAR\\Capabilities"
> "Firefox-308046B0AF4A39CB"="Software\\Clients\\StartMenuInternet\\Firefox-308046B0AF4A39CB\\Capabilities"
> "Print Studio"="SOFTWARE\\PrintStudio\\Capabilities"
> "Repetier-Host"="Software\\Repetier\\Capabilities"
> "VisualStudio.3bb3a044"="Software\\Microsoft\\VisualStudio_3bb3a044\\Capabilities"
> "Adobe Acrobat Reader DC"="Software\\Adobe\\Acrobat Reader\\DC\\Capabilities"
> "LibreOffice 6.0"="SOFTWARE\\The Document Foundation\\LibreOffice\\6.0\\Capabilities"

> [HKEY_CLASSES_ROOT\SearchDeflector]
>

**Comment by spikespaz** [comment]
What's strange is that the registry keys exist at all. I wouldn't expect that, if it can't write the default value.

When I created the binary distribution I included debug information so the stack trace would give me more information. Unfortunately, the debug files don't work on other systems apparently because there should have been a load more information in the crash report.

**Comment by spikespaz** [comment]
This may also be related to #10, something is preventing the program from accessing your registry.

**Comment by nicegamer7** [comment]
Yeah I was a little worried about the pre-compiled binaries, so I checked their sums against the source binaries where the code is provided, and they didn't match (hm). It's a nice all in one though, I just don't use it's binaries. Thanks for the warning.

Back on topic I guess.

[HKEY_CURRENT_USER\Software\Clients\SearchDeflector]
`it's blank :(`

[HKEY_LOCAL_MACHINE\SOFTWARE\Clients\SearchDeflector]
`it's blank :(`

[HKEY_LOCAL_MACHINE\SOFTWARE\RegisteredApplications]
`"File Explorer"="SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\Capabilities"
"Paint"="SOFTWARE\Microsoft\Windows\CurrentVersion\Applets\Paint\Capabilities"
"Windows Address Book"="Software\Clients\Contacts\Address Book\Capabilities"
"Windows Disc Image Burner"="Software\Microsoft\IsoBurn\Capabilities"
"Windows Photo Viewer"="Software\\Microsoft\\Windows Photo Viewer\Capabilities"
"Windows Search"="Software\Microsoft\Windows Search\Capabilities"
"Wordpad"="Software\Microsoft\Windows\CurrentVersion\Applets\Wordpad\Capabilities"
"Google Chrome"="Software\Clients\StartMenuInternet\Google Chrome\Capabilities"
"SpeedCrunch"="Software\SpeedCrunch\Capabilities"`

[HKEY_CLASSES_ROOT\SearchDeflector]
`it's blank :(`

**Comment by loginnotsecure** [comment]
@spikespaz Doesn't work, I just tried it. Here's the error report I got:

`I have encountered an error launching Search Deflector.
The error information follows below.

**File:** `std\windows\registry.d`

**Line:** 495

**Message:**
```
Value cannot be opened: "": The system cannot find the file specified. (error 2)
```

**Stack Trace:**
```
0x00BE5BAF
0x00BE5962
0x00BDEA3D

```
`

**Comment by nicegamer7** [comment]
I have encountered an error launching Search Deflector.
The error information follows below.

**File:** `std\windows\registry.d`

**Line:** 495

**Message:**
```
Value cannot be opened: "": The system cannot find the file specified. (error 2)
```

**Stack Trace:**
```
0x01185BAF
0x01185962
0x0117EA3D

```

That was with the one you sent above.

**Comment by spikespaz** [comment]
The issue may very likely be fixed in versions past 0.3.1. Download the latest installer [here](https://spikespaz.com/tools/repo-dl/?user=spikespaz&repo=search-deflector&file=SearchDeflector-Installer.exe).

When you run the installer, try the defaults first. Then, if it still doesn't work, select the "Register for all users" checkbox when you see it.

I moved almost all of the registry changing code to the installer itself in preparation for the Microsoft Store. This should handle your specific system quirks better than my code can.

If it still doesn't work after this I will provide free promotion codes for the store version of it once my package gets approved (should take a few days).

**Comment by nicegamer7** [comment]
I tried both with and without the new option, still not working.

**Comment by spikespaz** [comment]
For all of you having this issue, email me at promo@spikespaz.com, and link your GitHub profile in the body of the message. I will send a one-time use promotion code for the Microsoft Store version ([here](https://www.microsoft.com/store/productId/9P8ZJJ80RZ2K)) to compensate for your troubles. Additionally, you won't have this issue with the Store version.

**Comment by blackcomb-dev** [comment]
@spikespaz when I installed the Microsoft Store version, I got this error:

````I have encountered an error launching Search Deflector.
The error information follows below.

**File:** `std\windows\registry.d`

**Line:** 495

**Message:**
```
Value cannot be opened: "": The system cannot find the file specified. (error 2)
```

**Stack Trace:**
```
0x002831DF
0x00282F92
0x0027B0AD

``````

**Comment by spikespaz** [comment]
I don't understand! The store version doesn't even use the registry!

**Comment by blackcomb-dev** [comment]
@spikespaz after my computer finishes its Windows update I'll try again

**Comment by spikespaz** [comment]
@blackcomb-dev Don't forget to remove the old version (`C:/Program Files (x86)/Search Deflector/unins000.exe`) before using the Store version.

**Comment by blackcomb-dev** [comment]
@spikespaz just tried again after deleting old files, clearing cache, etc, and still gave me the same error

**Comment by blackcomb-dev** [comment]
@spikespaz 0.3.4 doesn't work

**Comment by spikespaz** [comment]
@blackcomb-dev You mean for this specific issue or does it just not work at all?

**Comment by blackcomb-dev** [comment]
@spikespaz sorry, I should have been more specific. Yes, this specific issue. I just updated from the Microsoft Store and it threw the same error

**Comment by spikespaz** [comment]
I have the exact same build of Windows on the slow ring. Is everyone here running preview? Sorry to send notifications to the watchers.

**Comment by Kwaadpepper** [comment]
Just to mention I tried the trial version on Store, and i still have this issue. I took care to uninstall other versions first. I have Windows Defender and Bitdefender. I do have EdgeDeflector Installed.

**Comment by spikespaz** [comment]
For some reason the issue is caused by the installer (newer versions) and the setup (older) failing to write the registry keys. That makes the deflector (or launcher) fail when it is later activated by the user.

I still don't know why the registry keys aren't present when needed, our original guess at Antivirus is probably the best one we have.

**Comment by Kwaadpepper** [comment]
As for more info Only version is showed before the crash. So reading available browsers make the try-catch to trigger an error.

**Comment by blackcomb-dev** [comment]
@spikespaz I only have Windows defender installed, no other anti-virus. But if you could provide a customized .reg file that we could try to fix this issue we can give you feedback etc

**Comment by Kwaadpepper** [comment]
May be a tip ? https://forum.dlang.org/post/bkdvcwzaaknaxwvepbqf@forum.dlang.org

Saw this ``` .getKey("Run", REGSAM.KEY_ALL_ACCESS);``` He say ```You need to use a REGSAM value (e.g. KEY_ALL_ACCESS) to open the key with write access:```

**Comment by Kwaadpepper** [comment]
Hey I think i found the issue : 

```string browserName = key.getValue("").value_SZ;``` on my end throw an unhandled exception because i got this on my registry ```[HKEY_LOCAL_MACHINE\SOFTWARE\Clients\StartMenuInternet\IEXPLORE.EXE]``` and it value_SZ is **undefined** on the registry.

**Comment by blackcomb-dev** [comment]
@spikespaz

**Comment by spikespaz** [comment]
Huh that would make a lot of sense. I had assumed that the issue occurred where it tried to write the settings. This isn't the case, it was trying to read the available browsers on your computers and that's where it was failing.

So maybe a try-catch is the solution? If the browser can't be read just skip it and let the user use the custom path option.

Try this version.

[SearchDeflector-Installer.zip](https://github.com/spikespaz/search-deflector/files/2562202/SearchDeflector-Installer.zip)

**Comment by spikespaz** [comment]
@Kwaadpepper Also, do you see this?

![image](https://user-images.githubusercontent.com/12502988/48206658-24c8b680-e32c-11e8-9401-2d356e5a89e0.png)

That `.value_SZ` is the getter for that value type, not the key. The key is `""`, an empty string because of the `(Default)` thing.

**Comment by Kwaadpepper** [comment]
![annotation 2018-11-08 160219](https://user-images.githubusercontent.com/6019313/48206900-b7d71000-e36f-11e8-978b-61dce94352b5.jpg)

This is what i have. Value as you see is "undefined".

**Comment by spikespaz** [comment]
@Kwaadpepper Yes the user interface will not change because it is part of the operating system.

[X](https://github.com/spikespaz/search-deflector/wiki/Troubleshooting#still-using-bing-in-the-search-ui)

**Comment by Kwaadpepper** [comment]
Ok nevermind. Thanks for the update. Program is working perfectly fine .

**Comment by blackcomb-dev** [comment]
> @Kwaadpepper The one I sent is working perfectly fine? The one compiled with the version 0.3.5 linked in an above comment?

I can confirm that it works perfect on my end

**Comment by blackcomb-dev** [comment]
@spikespaz can I close this issue now?

**Comment by spikespaz** [comment]
Resolved and confirmed by commit 6dd27d6.

Long awaited, this issue is closed.
